Presidio power skill (#1)
shiranr/azure-search-power-skills: Text/PresidioPIIAnonymization/README.md

# Text PII Anonymization Power Skill

[Presidio](https://github.com/microsoft/presidio) is an open-source tool to recognize, analyze and anonymize personally identifiable information (PII). Using trained ML models, Presidio was built to ensure sensitive text is properly managed and governed.

This Power Skill uses [Presidio](https://github.com/microsoft/presidio) Analyzer and Anonymizer
to find and remove PII entities. Even though Presidio supports several anonymization methods (hash, encrypt, redact, replace, mask), 
this Power Skill only uses redact and removes the PIIs completely from the text  .

> This skill is ideal for finding and removing PII entities from the search text.

Reviewer: Consider adding a disclaimer like we have on our readme, that we cannot guarantee 100% accuracy

Author: Done.

> ## Requirements

Reviewer: Are these requirements for installing locally or deploying?

Author: Done.

In addition to the common requirements described in the root [README.md](https://github.com/Azure-Samples/azure-search-power-skills/blob/master/README.md) 
file, this Power Skill requires spacy en_core_web_lg module being downloaded:
```python

> python -m spacy download en_core_web_lg

Reviewer: Is this the only requirement? What about presidio-analyzer and presidio-anonymizer? 
Also, it's not clear if these are the requirements to develop locally or to deploy the power skill into the user's cognitive search pipeline.

```

> To run this PowerSkill you will need:

Reviewer: Consider having this prior to the spacy stuff, as this is the basics for setting up a power skill

> * docker

Reviewer: ```suggestion
* Docker
```

* An Azure Blob storage container
* A provisioned Azure Cognitive Search (ACS) instance 
* A provisioned Azure Container Registry
* A Cognitive Services key in the region you deploy ACS to

Below is a full working example that you can get working E2E on sample data.

## How to implement

This section describes how to get this working on sample data and how it can be amended for your data.
 
1) ### Data

>    The first step is to view the sample data files here [train data](data/). 

Reviewer: ```suggestion
   The first step is to view the sample data. [Link to sample data](data/). 
```

> 1) ### Run the API

Reviewer: Consider splitting into Install and then Run

Author: I followed the summarization way of putting it down.

>    The next step is to run the API locally and test the model against a test record. Create a local python environment

Reviewer: If by API you mean the one that's exposed in this power skill, worth mentioning explicitly.

Author: Done.

   and install the requirements:
   
   ```python

>    python -m pip install -r requirements.txt

Reviewer: ```suggestion
   python -m pip install -r powerskill/requirements.txt
```

Reviewer: Why not add the spacy model installation here?

   ```
   Activate your environment and run the API locally, execute the following:
   ```python 
   python app.py
   ``` 
   Run the cell 
   [Test PII anonymization on our local running API](notebooks/PII%20Anonymization.ipynb#Test-our-text-on-our-local-running-API). 
   Make sure you rename the file [sample_env file to .env](powerskill/sample_env) and populate it with the relevant values. Use the
   variable ```bash URL_LOCAL``` as the URL.
1) ### Build the docker image 
   Now build the [docker image](Dockerfile) and upload the image to your container registry  

>    For this step you will need docker running so that we can build and test our inference API locally.

Reviewer: Inference API can be confusing. Consider being more explicit and call it "Anonymization API" or "Presidio inference API" or something similar.

Author: Done.

   You will also need a container registry for the build.

   Run the following command to build the inference API container image:

    ```bash
    docker build -t [container_registry_name.azurecr.io]/pii_anonymization:[your_tag] .  
    ```
    
    The container will require the following variables set at runtime, namely:
    
    ```bash

>     KEY=[YourSecretKeyCanBeAnything]    # This is a secret key - only requests with this key will be allowed

Reviewer: Is this mandatory for any powerskill? I'm not sure what this key does and how to use it.

Author: it's part of the authentication for the application. It is done in other powerskills as well, yes.

    
    Now we can test the container by running it locally with our variables:
    
    ```bash
    docker run -it --rm -p 5000:5000 -e DEBUG=true -e KEY=[YourSecretKeyCanBeAnything] 
    [container_registry_name.azurecr.io]/pii_anonymization:[your_tag]
    ```
    Upon starting, you will see the server initializing:
    ```bash
    INFO:     Uvicorn running on http://0.0.0.0:5000 (Press CTRL+C to quit)
    ```
    We are now ready to send a request. Run the cell 
    [Test PII anonymization on local running API](notebooks/PII%20Anonymization.ipynb#Test-our-text-on-our-local-running-API) to test
    the running container.
    
    The response should show the anonymized text. 
    
1) ### Deploy the container to an Azure Web App.

    We will deploy this as an [Azure App Service Web App](https://docs.microsoft.com/en-us/azure/app-service/configure-custom-container?pivots=container-linux).
    running a container.
    
    First we need to push our newly built image to our container registry.
    
    Run the following command:
    ```bash
    docker push [container_registry_name].azurecr.io/pii_anonymization:[your_tag]
    ```
    

>     In the [deployment folder](deployment/webapp) are two [terraform](https://www.terraform.io/)

Reviewer: ```suggestion
    In the [deployment folder](deployment/webapp) there are two [terraform](https://www.terraform.io/)
```

Author: Done.

    files to deploy the inference API to an App Service Web App for linux.
    
    The simplest is to open a [cloud shell in the portal](https://ms.portal.azure.com/#home) and upload
    the [main](deployment/webapp/main.tf) and [variables](deployment/webapp/variables.tf)
    to your cloud shell storage as this avoids the need for any installation. 
    
    Set the following values in the [main](deployment/webapp/main.tf) file:
    ```hcl-terraform
    backend "azurerm" {
        storage_account_name = "[your storage account name]"
        container_name = "[your storage container name]"
        key = "[your storage account key]"
        resource_group_name = "[your storage account resource group name]"
      }
    ```
    

>     Set the following values in the[variables](deployment/webapp/variables.tf)

Reviewer: ```suggestion
    Set the following values in the [variables](deployment/webapp/variables.tf)
```

Author: Done.
